File: taeho/baekjoon/python/2667.py
# N = int(input())
# graph = []
# visit = [[False] * N for _ in range(N)]
# iPos = [-1, +1, 0, 0]
# jPos = [0, 0, -1, 1]
# dngs = []
# dng = 1

# for _ in range(N):
#     graph.append(list(map(int, input().strip(" "))))


# def dfs(i, j):
#     global dng
#     visit[i][j] = True
#     for p in range(4):
#         newI, newJ = i + iPos[p], j + jPos[p]
#         if newI < 0 or newI >= N or newJ < 0 or newJ >= N:
#             continue
#         elif graph[newI][newJ] != 1 or visit[newI][newJ]:
#             continue
#         else:
#             dng += 1
#             dfs(newI, newJ)


# cnt = 0
# for i in range(N):
#     for j in range(N):
#         if graph[i][j] == 1 and visit[i][j] == False:
#             cnt += 1
#             dfs(i, j)
#             dngs.append(dng)
#             dng = 1
# print(cnt)
# for i in sorted(dngs):
#     print(i)

N = int(input())
graph = []
visit = [[False] * N for _ in range(N)]
dngs = []
dng = 0

for _ in range(N):
    graph.append(list(map(int, input().strip(" "))))


def dfs(i, j):
    global dng
    if i < 0 or i >= N or j < 0 or j >= N:
        return
    if visit[i][j] or graph[i][j] == 0:
        return
    visit[i][j] = True
    dng += 1
    dfs(i + 1, j)
    dfs(i - 1, j)
    dfs(i, j + 1)
    dfs(i, j - 1)


cnt = 0
for i in range(N):
    for j in range(N):
        if graph[i][j] == 1 and visit[i][j] == False:
            cnt += 1
            dfs(i, j)
            dngs.append(dng)
            dng = 0
print(cnt)
for i in sorted(dngs):
    print(i)

File: taeho/baekjoon/javascript/2667.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];
rl.on('line', line => {
  input.push(line);
}).on('close', () => {
  const N = +input.shift();
  const arr = input.map(e => Array.from(e, x => +x));
  const visit = Array.from(Array(N), _ => Array(N).fill(false));

  const xPos = [-1, 1, 0, 0];
  const yPos = [0, 0, -1, 1];
  const dngs = [];
  let dng = 0;
  let cnt = 0;

  const dfs = (x, y) => {
    visit[x][y] = true;
    dng++;
    for (let i = 0; i < 4; i++) {
      let newX = x + xPos[i];
      let newY = y + yPos[i];
      if (newX < 0 || newX >= N || newY < 0 || newY >= N) continue;
      else if (!arr[newX][newY] || visit[newX][newY]) continue;
      else {
        dfs(newX, newY);
      }
    }
  };

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (arr[i][j] && !visit[i][j]) {
        cnt++;
        dfs(i, j);
        dngs.push(dng);
        dng = 0;
      }
    }
  }
  console.log(cnt);
  dngs.sort((a, b) => a - b).forEach(e => console.log(e));
});
